sdk/python/feast/feature_store.py

```diff
@@ -425,24 +473,77 @@ def plan(
+            >>> registry_diff, infra_diff, new_infra = fs._plan(RepoContents({driver_hourly_stats_view}, set(), set(), {driver}, set())) # register entity and feature view
         """
         registry_diff = diff_between(
             self._registry, self.project, desired_repo_contents
         )
 
+        self._registry.refresh()
         current_infra_proto = (
             self._registry.cached_registry_proto.infra.__deepcopy__()
             if self._registry.cached_registry_proto
             else InfraProto()
         )
         desired_registry_proto = desired_repo_contents.to_registry_proto()
-        new_infra_proto = self._provider.plan_infra(
-            self.config, desired_registry_proto
-        ).to_proto()
+        new_infra = self._provider.plan_infra(self.config, desired_registry_proto)
+        new_infra_proto = new_infra.to_proto()
         infra_diff = diff_infra_protos(current_infra_proto, new_infra_proto)
 
-        return (registry_diff, infra_diff)
+        return (registry_diff, infra_diff, new_infra)
+
+    @log_exceptions_and_usage
+    def _apply_diffs(
+        self, registry_diff: RegistryDiff, infra_diff: InfraDiff, new_infra: Infra
+    ):
+        """Applies the given diffs to the metadata store and infrastructure.
+
+        Args:
+            registry_diff: The diff between the current registry and the desired registry.
+            infra_diff: The diff between the current infra and the desired infra.
+            new_infra: The desired infra.
+        """
+        entities_to_update = [
+            fco_diff.new_fco
+            for fco_diff in registry_diff.fco_diffs
+            if fco_diff.fco_type == FeastObjectType.ENTITY
+            and fco_diff.transition_type
+            in [TransitionType.CREATE, TransitionType.UPDATE]
+        ]
+        views_to_update = [
+            fco_diff.new_fco
+            for fco_diff in registry_diff.fco_diffs
+            if fco_diff.fco_type == FeastObjectType.FEATURE_VIEW
+            and fco_diff.transition_type
+            in [TransitionType.CREATE, TransitionType.UPDATE]
+        ]
+        odfvs_to_update = [
+            fco_diff.new_fco
+            for fco_diff in registry_diff.fco_diffs
+            if fco_diff.fco_type == FeastObjectType.ON_DEMAND_FEATURE_VIEW
+            and fco_diff.transition_type
+            in [TransitionType.CREATE, TransitionType.UPDATE]
+        ]
+        request_views_to_update = [
+            fco_diff.new_fco
+            for fco_diff in registry_diff.fco_diffs
+            if fco_diff.fco_type == FeastObjectType.REQUEST_FEATURE_VIEW
+            and fco_diff.transition_type
+            in [TransitionType.CREATE, TransitionType.UPDATE]
+        ]
+
+        # Validate all feature views and make inferences.
+        self._validate_all_feature_views(
+            views_to_update, odfvs_to_update, request_views_to_update
+        )
+        self._make_inferences(entities_to_update, views_to_update, odfvs_to_update)
```

Comment:
Add a TODO to move inference to be part of feast plan to avoid bugs with entityless FVs?